### ISSUE TITLE
Can not overlap content-type

### ISSUE THREAD

**Issue by kvokka** [apiaryio/dredd#1146]
**Describe the bug**

Given:
```
swagger: '2.0'
info:
  version: '0.1.0'
basePath: /api/v1
host: 'localhost:5000'
schemes:
  - http
consumes:
  - application/json
produces:
  - application/json; charset=utf-8

paths:
  /foos/{id}/:
    delete:
      description: 'foo'
      parameters:
        - name: "id"
          in: "path"
          description: "ID of foo"
          required: true
          type: integer
          enum:
          - 42
      responses:
        204:
          description: "OK"
```

Produce output:
```
expected:
headers:
    Content-Type: application/json; charset=utf-8

statusCode: 204


actual:
statusCode: 204
headers:
    cache-control: no-cache
    x-request-id: 03565c02-cdfd-49b9-af36-506cae55cb8a
    x-runtime: 0.841603
    connection: close
```

And i can not overlap it in the specification

**Expected behavior**
add the ability to add option to describe this in `yaml` files

**What's your `dredd --version` output?**

```
5.2.0
```

PS: I found the solution (and it works) to hack it, but i'm not sure, that it is semantically  right:

```
before "/api/v1/foos/{id}/ > Delete the foo > 204 > application/json; charset=utf-8" do |transaction|
  create :foo, id: 42
  transaction.fetch("expected", {}).fetch("headers", {}).delete("Content-Type")
end

```

**Comment by honzajavorek** [comment]
Hi @kvokka, I'm sorry, I must have missed this issue.

Do I understand correctly that you're trying to test 204 No Content response in a document where everything produces JSON by default, and Dredd considers this to be a problem, because the 204 No Content response should not have a Content-Type?

While this seems to be correct rigorously, I can see it can be annoying from user's perspective. @apiaryio/adt Do you think it would be a good idea to suppress Content-Type for 204 and 205 responses in the Swagger adapter?

**Comment by kvokka** [comment]
Hi @honzajavorek , than you for your responce!

Yes, you understood me right.

In this period I found one more, even more pretty solution
```
delete:
  tags:
    - Users
  description: User logout
  produces:
    - "text/plain"
```

And this construction allow to avoid this situation too. But, maybe it can be good idea to omit context-type check, if the body is empty?

**Comment by honzajavorek** [comment]
You are right, this is nicer solution. Maybe even this could work (did not try)?

```yaml
delete:
  tags:
    - Users
  description: User logout
  produces:
    - ""
```

I agree that it would be convenient to omit the content-type, but at the same time, the document should be universal and describe the API as it is, in the most real way. The semantics of OpenAPI 2 unfortunately lead to this, but omitting the content-type in the Swagger adapter or in Dredd means the API stays described with the problem in place for other Swagger tools you may use or integrate with.

In other words, if you have produces with JSON and 204 No Content response, other tools will/should still translate it to a response with a JSON content type and that's not how the API truly behaves, and Dredd wouldn't be able to warn you about it.

So at the end, I tend to think we should not introduce any specific behavior in this regard. Opinions?

**Comment by kvokka** [comment]
Thank you @honzajavorek !

I tried out your solution, and it is shorter and better. And it works :)

Semantically this idea looks even better, than separate flag. The only question is, that this feature is undocumented, but at the same time widely used. If you do not mind, I can add this idea to the docs in pr? IDK how it works with apiary, so is it ok, that the section will be only about swagger?

**Comment by honzajavorek** [comment]
@kvokka I'm sorry, I don't know why this always slips through my notifications 😞 Perhaps this section could contain an example of API descriptions with empty response bodies? https://dredd.org/en/latest/how-it-works.html#empty-response-body

**Comment by ehuebner** [comment]
Hi,
i am facing the same issue.
A delete is a perfect example:
In the success case it emits a status code 204, empty body and an empty content type is okay.
In a error case it emits a status code 4xx, with an error in the body and the content type needs to be application json.

Since i cannot define different content types per status code:
@honzajavorek What is your idea how to solve this problem?

We have been thinking about setting application json as a content type even for empty body requests. But RFC2616 said that is not correct.

**Comment by kylef** [comment]
Just to add to this, you don't need to set produces to an empty string, you can set produces to an empty array to represent that the operation produces no body with content-type. I think that's semantically the clearest as we are saying that the method produces no content types.

This is echoed in the Swagger 2.0 specification: 

> A list of MIME types the operation can produce. This overrides the produces definition at the Swagger Object. An empty value MAY be used to clear the global definition. Value MUST be as described under Mime Types.

From https://github.com/OAI/OpenAPI-Specification/blob/8a387ac71f7f31aaff5b910f1e0e99ad12ff3ef0/versions/2.0.md#operation-object

For example:

```yaml
swagger: '2.0'
info:
  title: 'HTTPBin'
  version: '1.0.0'
produces:
  - application/json
paths:
  /status/204:
    delete:
      description: 'delete status'
      produces: []
      responses:
        204:
          description: 'empty response'
```

```shell
$ dredd httpbin.yaml https://httpbin.org/
pass: DELETE (204) /status/204 duration: 382ms
complete: 1 passing, 0 failing, 0 errors, 0 skipped, 1 total
complete: Tests took 387ms
```

**Comment by ehuebner** [comment]
Hi @kylef.
Thanks for your answer.

Set a empty content type for an operation is clear. But lets take a simple example of the official examples: https://github.com/OAI/OpenAPI-Specification/blob/master/examples/v2.0/yaml/petstore-expanded.yaml

The response of 204 should have no content type because of the empty response body. But the 500 of the operation will have a response body and so it should have a content type. That is not possible to run in dredd because it expacts the same content type for all status codes of one operation. And openapi 2 let you only define one content type per operation.

Hoes your delete response also have an empty status code in an error case?

```
swagger: "2.0"
info:
  version: 1.0.0
consumes:
  - application/json
produces:
  - application/json
paths:
  /pets/{id}:
      delete:
        description: deletes a single pet based on the ID supplied
        operationId: deletePet
        parameters:
          - name: id
            in: path
            description: ID of pet to delete
            required: true
            type: integer
            format: int64
        responses:
          "204":
            description: pet deleted
          "500":
            description: unexpected error
            schema:
              $ref: '#/definitions/Error'
definitions:
  Error:
    required:
      - code
      - message
    properties:
      code:
        type: integer
        format: int32
      message:
        type: string
```

**Comment by kylef** [comment]
> Set a empty content type for an operation is clear. But lets take a simple example of the official examples: OAI/OpenAPI-Specification:examples/v2.0/yaml/petstore-expanded.yaml@master

> The response of 204 should have no content type because of the empty response body

I don't believe that is true based on how produces is described in the specification. However we can take the example you linked and compare it to other Swagger tools to see how it is handled. First, how does SwaggerUI render this?

<img width="528" alt="Screenshot 2019-05-08 at 17 41 51" src="https://user-images.githubusercontent.com/44164/57392577-1400dc80-71b9-11e9-9949-6e2978a8964a.png">

SwaggerUI shows the 204 response under the `application/json` content type.

SwaggerHub has a mock server, let's compare that:

```
$ curl http://virtserver.swaggerhub.com/kylef/petstore/1.0.0/pets/1 -X DELETE -i
HTTP/1.1 204 No Content
Content-Type: application/json
(remove other headers)
```

Both these existing tools use the `application/json` response content-type with 204 responses.

---

@ehuebner Correct me if I am wrong, but from my interpretation of the specification, OpenAPI 2 doesn't have a way to declare status codes with different produces media types. This is also a problem if you have a specific content-type for error cases, such as `application/problem+json` and you use `application/json` for success types.

This is resolved in OpenAPI 3.0 (see bottom of my comment for alternative version in OpenAPI 3.0):

Take the following example:

```yaml
swagger: "2.0"
info:
  title: Example
  version: 1.0.0
produces:
  - application/json
  - application/problem+json
paths:
  /example:
    get:
      description: example, which randomly fails or not
      responses:
        '200':
          description: success
          schema:
            type: object
        '400':
          description: an error
          schema:
            type: object
```

In OpenAPI 3 this was addressed and the API can be described as:

```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: Example
paths:
  /example:
    get:
      responses:
        '200':
          description: success
          content:
            application/json:    
              schema:
                type: object
        '400':
          description: failure
          content:
            application/problem+json:    
              schema:
                type: object
```

**Comment by honzajavorek** [comment]
Thanks @kylef for answering here. I think this is a limitation of the OpenAPI 2 format and Dredd cannot do much about it, given there are at least two other formats (OpenAPI 3, API Blueprint) not having the same issue. I'd say adjusting the content type by Dredd hooks is a valid solution in such case, as it is a special case, and those are something hooks are exactly for.

I think this could go to the docs about empty bodies though. @kvokka @ehuebner would you be willing to contribute a sentence to Dredd's docs about this?
